code/modules/research/anomaly/anomaly_refinery.dm

```diff
@@ -0,0 +1,300 @@
+	var/light = arguments[EXARG_KEY_LIGHT_RANGE]
+	var/explosion_range = max(heavy, medium, light, 0)
+	var/required_range = get_required_radius(inserted_core.anomaly_type)
+	var/turf/location = get_turf(src)
+
+	var/cap_multiplier = SSmapping.level_trait(location.z, ZTRAIT_BOMBCAP_MULTIPLIER)
+	if(isnull(cap_multiplier))
+		cap_multiplier = 1
+	var/capped_heavy = min(GLOB.MAX_EX_DEVESTATION_RANGE * cap_multiplier, heavy)
+	var/capped_medium = min(GLOB.MAX_EX_HEAVY_RANGE * cap_multiplier, medium)
+	SSexplosions.shake_the_room(location, explosion_range, (capped_heavy * 15) + (capped_medium * 20), capped_heavy, capped_medium)
+
+	if(explosion_range < required_range)
+		test_status = "Resultant detonation failed to produce enough implosive power to compress [inserted_core]. Items ejected."
+		return
+
+	if(test_status)
+		return
+	inserted_core = inserted_core.create_core(src, TRUE, TRUE)
+	test_status = "Success. Resultant detonation has theoretical range of [explosion_range]. Required radius was [required_range]. Core production complete."
+	return
+
+/**
+ * Handles timing out the test after a while.
+ */
+/obj/machinery/research/anomaly_refinery/proc/timeout_test()
+	timeout_timer = null
+	if(!test_status)
+		test_status = "Transfer valve resulted in negligible explosive power. Items ejected."
+	end_test(test_status)
+
+/// This is not the real valve opening process. This is the simulated one used for displaying reactions.
+/obj/machinery/research/anomaly_refinery/proc/simulate_valve()
+	if(!inserted_bomb?.tank_one || !inserted_bomb?.tank_two)
+		eject_bomb()
+		return FALSE
+
+	if(reaction_increment == 0)
+		var/datum/gas_mixture/first_gasmix = inserted_bomb.tank_one.return_air()
+		var/datum/gas_mixture/second_gasmix = inserted_bomb.tank_two.return_air()
+
+		combined_gasmix = new(70)
+		combined_gasmix.volume = first_gasmix.volume + second_gasmix.volume
+		combined_gasmix.merge(first_gasmix.copy())
+		combined_gasmix.merge(second_gasmix.copy())
+	else
+		combined_gasmix.react()
+	
+	reaction_increment += 1
+
+/// We dont allow incomplete valves to go in but do code in checks for incomplete valves. Just in case.
+/obj/machinery/research/anomaly_refinery/proc/eject_bomb(mob/user)
+	if(!inserted_bomb || active)
+		return
+	if(user)
+		user.put_in_hands(inserted_bomb)
+		to_chat(user, span_notice("You remove [inserted_bomb] from [src]."))
+	else
+		inserted_bomb.forceMove(drop_location())
+	QDEL_NULL(combined_gasmix)
+	reaction_increment = 0
+
+/obj/machinery/research/anomaly_refinery/proc/eject_core(mob/user)
+	if(!inserted_core || active)
+		return
+	if(user)
+		user.put_in_hands(inserted_core)
+		to_chat(user, span_notice("You remove [inserted_core] from [src]."))
+	else
+		inserted_core.forceMove(drop_location())
+
+/// We rely on exited to clear references.
+/obj/machinery/research/anomaly_refinery/Exited(atom/movable/gone, direction)
+	if(gone == inserted_bomb)
+		inserted_bomb = null
+		tank_to_target = null
```

Comment:
I think you might want a check for tank_to_target here

Reply:
`tank_to_target` should always be one of the ttv's tank. I think that gone will never be tank_to_target since the tanks are inside ttv's loc.